<a href="https://colab.research.google.com/github/Aissa-C/Anal-tica/blob/main/Reto_1_NB_Aissa_Ceballos.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

**Reto 1. Modelo Naive Bayes**

0. Cargue de las librerías de trabajo

In [ ]:
import numpy as np
import pandas as pd
from sklearn.naive_bayes import GaussianNB
from sklearn.metrics import confusion_matrix

1. Cargue de datos

In [ ]:
nxl= '/content/1. BD1_Longitudinal Survey_Int.xlsx'
XDB= pd.read_excel(nxl, sheet_name=0)
XDB=XDB.dropna()

#Selección de las variables de interés
XDB=XDB[['Age', 'Mother_Edu', 'Father_Edu', 'Height', 'Weight', 'Income',
          'Self_Esteem', 'FamilySize', 'Urban']]

#Datos de entrada
XD= np.array(XDB[['Age', 'Mother_Edu', 'Father_Edu', 'Height', 'Weight',
                  'Income', 'Self_Esteem', 'FamilySize']])
#Datos de salida
yd= np.array(XDB[['Urban']])

2. Implementación del modelo

In [ ]:
mnb= GaussianNB()
mnb.fit(XD,yd)
ydp= mnb.predict(XD)

#Medias de las variables para cada categoría
u= mnb.theta_
sigma= np.sqrt(mnb.var_)

#Total de datos y % de datos por categoría
Long_datos= len(XD)
print("El número total de datos de la base es:", Long_datos)

nd= mnb.class_count_
print("El número de datos por categoría es:", nd)

ppd= nd/Long_datos
print("Los porcentajes de datos por categoría de clasificación son:", ppd)

#Variaciones/límites
LI= u-sigma
LS= u+sigma

print("La media de las variables por categoría es:")
print(u)

print("El límite inferior de la variación por categoría es:")
print(LI)

print("El límite superior de la variación por categoría es:")
print(LS)

#Métrica de evaluación
cm= confusion_matrix(yd,ydp)

El número total de datos de la base es: 5683
El número de datos por categoría es: [1173. 4510.]
Los porcentajes de datos por categoría de clasificación son: [0.20640507 0.79359493]
La media de las variables por categoría es:
[[1.83810742e+01 1.07612958e+01 1.03324808e+01 6.72148338e+01
  1.48857630e+02 3.03672506e+04 2.20272805e+01 4.74168798e+00]
 [1.83569845e+01 1.11769401e+01 1.12565410e+01 6.69031042e+01
  1.44760976e+02 3.34278929e+04 2.26004435e+01 4.68780488e+00]]
El límite inferior de la variación por categoría es:
[[ 1.58996876e+01  7.58892579e+00  6.46311334e+00  6.31369799e+01
   1.18746964e+02 -1.33465823e+03  1.78344685e+01  2.16621834e+00]
 [ 1.58811112e+01  7.79013864e+00  7.15097738e+00  6.26897933e+01
   1.14669654e+02 -2.69533531e+03  1.84154526e+01  2.24996381e+00]]
El límite superior de la variación por categoría es:
[[2.08624608e+01 1.39336659e+01 1.42018483e+01 7.12926877e+01
  1.78968296e+02 6.20691595e+04 2.62200925e+01 7.31715762e+00]
 [2.08328577e+01 1.4563741

/usr/local/lib/python3.10/dist-packages/sklearn/utils/validation.py:1183: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)


3. Evaluación del modelo

In [ ]:
VN= cm[0,0]
FP= cm[0,1]
FN= cm[1,0]
VP= cm[1,1]


Exactitud= (VP+VN)/len(XD)
print("La exactitud es:", Exactitud)

TE= (FP+FN)/len(XD)
print("La tasa de error es:", TE)

Sensibilidad= VP/(VP+FP)
print("La sensibilidad es:", Sensibilidad)

Especificidad= VN/(VN+FN)
print("La especificidad es:", Especificidad)

Precision= VP/(VP+FN)
print("La precisión es:", Precision)

PredNeg= VN/(VN+FP)
print("La probabilidad de negativo es:", PredNeg)

La exactitud es: 0.7932430054548654
La tasa de error es: 0.2067569945451346
La sensibilidad es: 0.793729082261758
La especificidad es: 0.3333333333333333
La precisión es: 0.9991130820399113
La probabilidad de negativo es: 0.0017050298380221654


4. Qué pasa con una sola persona

In [ ]:
Xun= np.array([[34, 9, 6, 65, 135, 50000, 30, 3]])
ydun= mnb.predict(Xun)
print(ydun)

[1.]


**Análisis de los resultados**
Gracias a los resultados arrojados por el modelo podemos observar que:
1. Las personas del área urbana presentan educación de la madre(11.17), educación del padre(11.25), ingresos(33,427.89), y autoestima(22.6) mayores a comparación de las personas del área rural; pero, presentan edades(18.35), alturas(66.9), pesos(144.76) y tamaños de familia(4.68) menores a las personas del área rural.

2. El modelo funcionó mejor clasificando a los positivos(área urbana), como es evidenciable gracias al indicador de sensibilidad con un 79.37% de certeza.

3. En términos generales, el modelo logró una exactitud por encima del 79.32% en promedio, y una tasa de error cercana al 20.67% en promedio. Debido a la cifra tan alta de exactitud que obtuvo, esto lo convierte en un buen modelo.

4. De acuerdo con el modelo, la persona con estas características ([[34, 9, 6, 65, 135, 50000, 30, 3]]), vive en el área urbana.